lib/vector-core/src/stream/batcher.rs

```diff
@@ -1,742 +1,261 @@
+        let this = self.project();
+        *this.current_size = 0;
+        std::mem::take(this.batch)
     }
 
-    impl TestTimer {
-        fn new(responses: Vec<Poll<Option<u8>>>) -> Self {
-            Self {
-                responses,
-                valid_keys: HashSet::new(),
-            }
-        }
+    fn push_item(self: Pin<&mut Self>, item: S::Item, item_size: usize) {
+        let this = self.project();
+        this.batch.push(item);
+        *this.current_size += item_size;
     }
 
-    impl KeyedTimer<u8> for TestTimer {
-        fn clear(&mut self) {
-            self.valid_keys.clear();
+    fn reset_timer(self: Pin<&mut Self>) {
+        let timeout = self.timeout;
+        let mut this = self.project();
+        if let MaybeProj::Some(timer) = this.timer.as_mut().project() {
+            timer.reset(tokio::time::Instant::now() + timeout);
+        } else {
+            this.timer.set(Maybe::Some(tokio::time::sleep(timeout)));
         }
+    }
+}
 
-        fn insert(&mut self, item_key: u8) {
-            self.valid_keys.insert(item_key);
-        }
+impl<S, I> Stream for Batcher<S, I>
+where
+    S: Stream,
+    I: ItemBatchSize<S::Item>,
+{
+    type Item = Vec<S::Item>;
 
-        fn poll_expired(&mut self, _cx: &mut Context) -> Poll<Option<u8>> {
-            match self.responses.pop() {
-                Some(Poll::Pending) => unreachable!(),
-                None | Some(Poll::Ready(None)) => Poll::Ready(None),
-                Some(Poll::Ready(Some(k))) => {
-                    if self.valid_keys.contains(&k) {
-                        Poll::Ready(Some(k))
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            match self.as_mut().project().stream.poll_next(cx) {
+                Poll::Ready(None) => {
+                    if self.batch.is_empty() {
+                        return Poll::Ready(None);
                     } else {
-                        Poll::Ready(None)
+                        return Poll::Ready(Some(self.take_batch()));
                     }
                 }
-            }
-        }
-    }
-
-    fn arb_timer() -> impl Strategy<Value = TestTimer> {
-        // The timer always returns a `Poll::Ready` and never a `Poll::Pending`.
-        Vec::<(bool, u8)>::arbitrary()
-            .prop_map(|v| {
-                v.into_iter()
-                    .map(|(b, k)| {
-                        if b {
-                            Poll::Ready(Some(k))
-                        } else {
-                            Poll::Ready(None)
+                Poll::Ready(Some(item)) => {
+                    let item_size = self.item_batch_size.size(&item);
+                    if self.size_fits_in_batch(item_size) {
+                        self.as_mut().push_item(item, item_size);
+                        if self.is_batch_full() {
+                            return Poll::Ready(Some(self.as_mut().take_batch()));
```

Comment:
It seems like we should also drop the timer when we take a batch, since we always reset it when we add the first item back to the batch.